acceptancetests/jujupy/stream_server.py

```diff
@@ -0,0 +1,293 @@
+    :param series: String series to use instead of that of the passed binary.
+      Allows one to overwrite the series of the juju client.
+    :returns: String path to generated
+    """
+    _series_lookup = dict(
+        artful='ubuntu',
+        bionic='ubuntu',
+        trusty='ubuntu',
+        xenial='ubuntu',
+        win2012='win2012',
+        win2012hv='win2012',
+        win2012hvr2='win2012',
+        win2012r2='win2012',
+        win2016='win2012',
+        win2016nano='win2012',
+        win7='win2012',
+        win81='win2012',
+        win8='win2012',
+        win10='win2012',
+    )
+
+    bin_dir = os.path.dirname(juju_bin_path)
+    try:
+        jujud_path = os.path.join(
+            bin_dir,
+            [f for f in os.listdir(bin_dir) if f == 'jujud'][0])
+    except IndexError:
+        raise RuntimeError('Unable to find jujud binary in {}'.format(bin_dir))
+
+    try:
+        version_output = subprocess.check_output(
+            [jujud_path, 'version']).rstrip('\n')
+        version, bin_series, arch = version_output.split('-')
+        bin_agent_series = _series_lookup.get(series, series)
+    except subprocess.CalledProcessError as e:
+        raise RuntimeError(
+            'Unable to query jujud for version details: {}'.format(e))
+    except IndexError:
+        raise RuntimeError(
+            'Unable to determine version, series and arch from version '
+            'string: {}'.format(version_output))
+
+    agent_tgz_name = 'juju-{version}-{series}-{arch}.tgz'.format(
+        version=version,
+        series=series if series else bin_agent_series,
+        arch=arch
+    )
+
+    log.debug('Creating agent file: {}'.format(agent_tgz_name))
+    tgz_path = os.path.join(tmp_dir, agent_tgz_name)
+    with tarfile.open(tgz_path, 'w:gz') as tar:
+        tar.add(jujud_path, arcname='jujud')
+
+    return tgz_path
+
+
+def _generate_product_json(content_id, version, arch, series, agent_tgz_path):
+    """Return dict containing product metadata from provided args."""
+    series_name, series_code = _get_series_details(series)
+    tgz_name = os.path.basename(agent_tgz_path)
+    file_details = _get_tgz_file_details(agent_tgz_path)
+    item_name = '{version}-{series}-{arch}'.format(
+        version=version,
+        series='{}:{}'.format(series_name, series_code),
+        arch=arch)
+    return dict(
+        arch=arch,
+        content_id='com.ubuntu.juju:{}:tools'.format(content_id),
+        format='products:1.0',
+        ftype='tar.gz',
+        item_name=item_name,
+        md5=file_details['md5'],
+        path=os.path.join('agent', tgz_name),
+        product_name='com.ubuntu.juju:{series_code}:{arch}'.format(
+            series_code=series_code,
+            arch=arch),
+        release=series_name,
+        sha256=file_details['sha256'],
+        size=file_details['size'],
+        version=version,
+        version_name=datetime.utcnow().strftime('%Y%m%d')
+    )
+
+
+_series_lookup = dict(
```

Comment:
same thing here about avoiding hardcoding. maybe we use an algo to oscillate between 04 and 10, and count up? So trusty starts with t, and 14.04. You could extrapolate from there right?